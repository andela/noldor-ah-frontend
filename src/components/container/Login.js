import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import propTypes from 'prop-types';
import axiosRequest from '../../utilities/axiosRequest';
import saveLoginDetails from '../../actions/loginAction';
import {
  Input, Notification, Button,
} from '../presentational/index';
import './styles/Login.scss';

const mapDispatchToProps = dispatch => ({
  storeLoginDetails: details => dispatch(saveLoginDetails(details))
});

class ConnectedLogin extends Component {
  constructor(props) {
    super(props);

    this.state = {
      email: '',
      password: '',
      status: '',
      message: '',
      display: 'none',
    };
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleInputChange = e => this.setState({ [e.target.name]: e.target.value });

  handleSubmit = (e) => {
    e.preventDefault();

    const { history, storeLoginDetails } = this.props;
    const {
      email, password,
    } = this.state;
    const user = { email, password };

    return axiosRequest.post('/users/login', user)
      .then((response) => {
        const { token, id: userId, message: resMessage } = response.data;
        storeLoginDetails({ token, userId });
        this.setState({ message: resMessage, display: 'block', status: 'success' });
        setTimeout(() => history.push('profile', { prev: 'login' }), 500);
      })
      .catch((err) => {
        if (err.response.status === 401) {
          return this.setState({
            message: err.response.data.error.email,
            status: 'error',
            display: 'block',
          });
        }
        return this.setState({
          message: err.response.data.message,
          status: 'error',
          display: 'block',
        });
      });
  }

  render() {
    const {
      email, password,
      status, message, display
    } = this.state;

    return (
      <div className="container columns is-desktop level login-container">
        <div className="container column is-half login-left" />
        <div className="container column level level-item login-right">
          <div className="login-intro-container">
            <h2 className="level-item login-intro">WELCOME BACK</h2>
            <p className="level-item login-intro">Login to your account below</p>
          </div>
          <form className="login-form" onSubmit={this.handleSubmit}>
            <Notification
              status={status}
              message={message}
              display={display}
            />
            <Input
              title="E-mail"
              name="email"
              type="email"
              value={email}
              onChange={this.handleInputChange}
              placeholder="Enter your email address here"
              required
            />

            <Input
              title="Password"
              name="password"
              type="password"
              value={password}
              onChange={this.handleInputChange}
              placeholder="Enter your password here"
              required
            />

            <Button
              type="submit"
              text="LOGIN"
            />
          </form>
          <p
            className="cta"
          >
          No account yet? <Link to="signup">Sign up here</Link></p>
          <p
            className="cta"
          >
          Forgot password? <Link to="forgot-password">click here</Link></p>
        </div>
      </div>
    );
  }
}

ConnectedLogin.propTypes = {
  history: propTypes.object.isRequired,
  storeLoginDetails: propTypes.func.isRequired,
};

const Login = connect(null, mapDispatchToProps)(ConnectedLogin);

export default Login;
