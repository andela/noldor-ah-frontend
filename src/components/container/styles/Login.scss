@import '../../../utilities//scss-exports/variables';

.login-left {
  height: 600px;
  width: 500px;
  background-image: url('../../../static/images/login-img.jpg');
  background-position: center;
  background-color: #F4F4F4;
}

.login-container {
  justify-items: center;
  margin: 5% auto;
  max-width: 80%;
  background-color: #fff;
  box-shadow: 2px 2px 2px 2px rgba(0, 0, 0, 0.10);
  font-size: 1em;

  .login-intro-container {
    margin-bottom: 1.5em;
  }

  .login-right {
    padding: 2em;
  }

  input::placeholder {
    font-size: .8em;
  }

    h2 {
      font-weight: bold;
    }
}

.login-container {
  .form-group {
    display: block;
    margin-bottom: 2em;
  }
}

.login-intro:first-child, .login-container {
   .form-label {
    font-weight: bold;
   }
}

@media only screen and (max-width: $ipad) {
  html {
    background-color: #fff;
  }

  .login-container {
    width: 90%;
    box-shadow: none;

    .login-intro-container {
      margin-bottom: 10%;
    }

    .login-left {
      display: none;
    }

    .form-button {
      width: 100%;
    }
  }
}

.cta {
  margin-bottom: -.5em;
  margin-top: .8em;
  font-size: .8em;
}

.cta a {
  text-decoration: none;
  color: $ah-primary;
  font-size: .8em;
}

.cta a:hover {
  border-bottom: 1px solid $ah-grey;
  padding-bottom: 0.2em;
}
