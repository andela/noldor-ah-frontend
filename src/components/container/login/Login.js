import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import propTypes from 'prop-types';
import { loginRequest } from '../../../redux/actions/login/loginAction';
import { Input, Notification, Button } from '../../presentational/index';
import { LOGIN_SUCCESS } from '../../../redux/types/login';
import '../styles/Login.scss';

const mapStateToProps = state => ({
  notification: state.notification,
});

const mapDispatchToProps = dispatch => ({
  login: user => dispatch(loginRequest(user))
});

export class ConnectedLogin extends Component {
  constructor(props) {
    super(props);

    this.state = {
      email: '',
      password: '',
      status: '',
      message: '',
      display: 'none',
    };
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleInputChange = e => this.setState({ [e.target.name]: e.target.value });

  handleSubmit = async (e) => {
    e.preventDefault();

    const {
      history, login,
    } = this.props;

    try {
      const response = await login({ ...this.state });
      this.setState({
        display: 'block',
        message: this.props.notification.message,
        status: this.props.notification.type,
      });
      if (response === LOGIN_SUCCESS) {
        return setTimeout(() => history.push('/', { prev: 'login' }), 500);
      }
    } catch (error) {
      /* do nothing */
    }
  }

  render() {
    const {
      email, password,
      status, message, display
    } = this.state;

    return (
      <div className="container is-desktop level login-container">
        <div className="container column is-half login-left" />
        <div className="container column level level-item login-right">
          <div className="login-intro-container">
            <h2 className="level-item login-intro">WELCOME BACK</h2>
            <p className="level-item login-intro">Login to your account below</p>
          </div>
          <form className="login-form" onSubmit={this.handleSubmit}>
            <Notification
              status={status}
              message={message}
              display={display}
            />
            <Input
              id="email"
              title="E-mail"
              name="email"
              type="email"
              value={email}
              onChange={this.handleInputChange}
              placeholder="Enter your email address here"
              required
            />

            <Input
              id="password"
              title="Password"
              name="password"
              type="password"
              value={password}
              onChange={this.handleInputChange}
              placeholder="Enter your password here"
              required
            />

            <Button
              type="submit"
              text="LOGIN"
            />
          </form>
          <p
            className="cta"
          >
          No account yet? <Link to="signup">Sign up here</Link></p>
          <p
            className="cta"
          >
          Forgot password? <Link to="forgot-password">click here</Link></p>
        </div>
      </div>
    );
  }
}

ConnectedLogin.propTypes = {
  history: propTypes.object,
  dispatch: propTypes.func,
  notification: propTypes.object,
};

const Login = connect(mapStateToProps, mapDispatchToProps)(ConnectedLogin);

export default Login;
