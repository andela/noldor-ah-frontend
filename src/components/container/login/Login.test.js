import React from 'react';
import { shallow } from 'enzyme';
import { ConnectedLogin } from './Login';

describe('<Login /> shallow rendering tests', () => {
  const mockLoginFn = jest.fn();
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ConnectedLogin login={mockLoginFn} />);
  });

  it('should render Login successfully', () => {
    expect(wrapper.find('.login-form').exists()).toBe(true);
    expect(wrapper.find('.login-intro').length).toBe(2);
    expect(wrapper.find('.login-intro-container')
      .childAt(1).text()).toBe('Login to your account below');
  });

  it('matches the snapshot', () => {
    const tree = shallow(<ConnectedLogin />);
    expect(tree).toMatchSnapshot();
  });

  it('should render all the children components successfully', () => {
    expect(wrapper.find('Notification').exists()).toBe(true);
    expect(wrapper.find('Input').exists()).toBe(true);
    expect(wrapper.find('Button').exists()).toBe(true);
  });

  it('should handle input changes successfully', () => {
    wrapper.find('#email').simulate(
      'change',
      {
        target: {
          name: 'email',
          value: 'papa@smurfs.com'
        }
      }
    );

    wrapper.find('#password').simulate(
      'change',
      {
        target: {
          name: 'password',
          value: 'GargamelIsComing!'
        }
      }
    );

    wrapper.find('form').simulate('submit', { preventDefault() {} });
    expect(mockLoginFn.mock.calls[0][0]).toEqual({
      email: 'papa@smurfs.com',
      password: 'GargamelIsComing!',
      display: 'none', // because state also handles other stuff
      message: '',
      status: ''
    });
  });

  it('should submit successfully', () => {
    wrapper.find('form').simulate('submit', { preventDefault() {} });
    expect(mockLoginFn.mock.calls.length).toBe(1);
  });
});
