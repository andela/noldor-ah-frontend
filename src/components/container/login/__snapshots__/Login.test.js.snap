// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> shallow rendering tests matches the snapshot 1`] = `
<div
  className="container columns is-desktop level login-container"
>
  <div
    className="container column is-half login-left"
  />
  <div
    className="container column level level-item login-right"
  >
    <div
      className="login-intro-container"
    >
      <h2
        className="level-item login-intro"
      >
        WELCOME BACK
      </h2>
      <p
        className="level-item login-intro"
      >
        Login to your account below
      </p>
    </div>
    <form
      className="login-form"
      onSubmit={[Function]}
    >
      <Notification
        display="none"
        message=""
        status=""
      />
      <Input
        className="form-control"
        id="email"
        name="email"
        onChange={[Function]}
        pattern={null}
        placeholder="Enter your email address here"
        required={true}
        title="E-mail"
        type="email"
        value=""
      />
      <Input
        className="form-control"
        id="password"
        name="password"
        onChange={[Function]}
        pattern={null}
        placeholder="Enter your password here"
        required={true}
        title="Password"
        type="password"
        value=""
      />
      <Button
        className="form-button"
        text="LOGIN"
        type="submit"
      />
    </form>
    <p
      className="cta"
    >
      No account yet? 
      <Link
        replace={false}
        to="signup"
      >
        Sign up here
      </Link>
    </p>
    <p
      className="cta"
    >
      Forgot password? 
      <Link
        replace={false}
        to="forgot-password"
      >
        click here
      </Link>
    </p>
  </div>
</div>
`;
